File: libs/power-log-parser/src/lib/cards/raza-the-resealed.ts
import { GameTag } from '@firestone-hs/reference-data';
import { Action } from '../models/action';
import { FullEntity } from '../models/entity';
import { Node } from '../models/node';
import type { GameState } from '../state/game-state';
import type { StateFacade } from '../state/state-facade';

export class RazaTheResealed {
	static PredictCardId(
		gameState: GameState,
		creatorCardId: string,
		creatorEntityId: number,
		node: Node,
		stateFacade: StateFacade | null,
	): string | null {
		if (node.Parent?.Type !== Action) {
			return null;
		}

		const act = node.Parent.Object as Action;
		const actionEntity = gameState.CurrentEntities.get(act.Entity);
		if (!actionEntity) {
			return null;
		}

		if (actionEntity.CardIdsToCreate.length === 0) {
			const subSpell = (node.Object as FullEntity).SubSpellInEffect;
			if (subSpell?.Targets != null) {
				actionEntity.CardIdsToCreate = subSpell.Targets.map(
					(t) => gameState.CurrentEntities.get(t)?.CardId ?? '',
				);
			}
		}

		if (actionEntity.CardIdsToCreate.length > 0) {
			const result = actionEntity.CardIdsToCreate[0];
			actionEntity.CardIdsToCreate.splice(0, 1);
			return result;
		}
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Cards/RazaTheResealed.cs
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using HearthstoneReplays.Parser;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;
using HearthstoneReplays.Parser.ReplayData.Entities;

namespace HearthstoneReplays.Events.Cards
{
    internal class RazaTheResealed
    {
        public static string PredctCardId(GameState gameState, string creatorCardId, int creatorEntityId, Node node, StateFacade stateFacade)
        {
            if (node.Parent?.Type != typeof(Action))
            {
                return null;
            }

            var act = node.Parent.Object as Action;
            var actionEntity = gameState.CurrentEntities.GetValueOrDefault(act.Entity);
            if (actionEntity.CardIdsToCreate.Count == 0)
            {
                var subSpell = (node.Object as FullEntity).SubSpellInEffect;
                // No targets when the opponent plays it
                if (subSpell?.Targets != null)
                {
                    actionEntity.CardIdsToCreate = subSpell.Targets.Select(t => gameState.CurrentEntities.GetValueOrDefault(t)?.CardId).ToList();
                }
            }

            if (actionEntity.CardIdsToCreate.Count > 0)
            {
                var result = actionEntity.CardIdsToCreate[0];
                actionEntity.CardIdsToCreate.RemoveAt(0);
                return result;
            }
            return null;
        }
    }
}
